[4~9단계 - 방탈출 예약 관리] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 지노! 
아직 부족한 부분이 많아 코드에 많이 드러날 것 같아요!
step2 미션도 잘 부탁드립니다 🙇‍♂️
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
- [셀프체크](https://prolog.techcourse.co.kr/studylogs/4229)
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
step2는 객체지향 생활체조 원칙보단 스프링에 대해서 조금 더 집중하여 진행했습니다. db와 상호작용 하다보니 엔티티와 도메인의 구분이 모호해짐을 느꼈습니다. 그래서 도메인 객체에 대해서 어떻게 하면 객체지향 생활체조 원칙을 잘 지킬 수 있는지 헷갈렸습니다. 
이외에 원칙을 지킨부분도 보인다고 생각했기에 4점으로 정했습니다!

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

## Dto 의 패키지 위치
Layered Architecture로 계층화를 하면서 dto와 객체의 변환 로직을 기존의 Controller 에서 Service로 이동하게 되었습니다.

그래서 원래 controller 패키지에 존재하던 dto를 service 패키지로 이동하게 되었는데 이유는 다음과 같습니다.

- Layered Architecture 의 목적은 계층화이고, 상위 계층은 하위 계층에 의존해야하며, 하위 계층은 상위 계층에 의존하면 안된다.
왜냐하면 상위 - 하위 계층이 서로 의존하게 되면 순환참조 문제가 생길수가 있고, 의존성의 방향이 양방향으로 된다.
- 계층화의 목적에 따라서 servie에서 dto를 변환해주는 것은 servie에서 dto를 의존하는 것이므로 dto의 위치도 controller에서 service로 옮기는 것이 좋다고 판단하였음.

하지만 dto를 어느 layer의 하위에 두지 않고, 독립적으로 위치시킬수도 있다고 생각합니다. 하지만 저는 dto는 현재 service에서 직접적으로 사용됨 + 위와같은 이유로 service 패키지 내부에 위치시켰습니다. 
이 선택에 대한 지노의 의견이 궁금합니다!

미션을 진행하면서 의식하지 못한 문제점들이 많을 것 같다고 생각이들어요. 의식하지 못한 부분에 대해서 피드백 해주시면 감사하겠습니다!

나머지 코드적은 부분은 코멘트에 남겨놓았습니다 : )

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

dye0p committed:
chore: 사용하지 않는 의존성 제거

dye0p committed:
chore: h2 콘솔 기능 활성화

dye0p committed:
feat: 데이터베이스 스키마 추가

dye0p committed:
test: DataSource 접근 테스트 추가

dye0p committed:
test: 합쳐진 테스트 분리

dye0p committed:
refactor: 데이터베이스를 이용하여 예약을 조회 할 수 있도록 변경

dye0p committed:
test: 테스트 네이밍에 단계 표시

dye0p committed:
refactor: 데이터베이스를 이용하여 예약을 추가/삭제 할 수 있도록 변경

dye0p committed:
refactor: insert로직을 SimpleJdbcInsert 객체를 사용하도록 변경

- JdbcTemplate.update() 보다 간단하게 insert 할 수 있다.

dye0p committed:
refactor: Reservation/Time 버튼을 클릭 했을 때 이동하는 URL 변경

- /reservation -> /admin/reservation
- /time -> /admin/time

dye0p committed:
feat: 시간 조회 페이지 API 추가

dye0p committed:
feat: reservation_time 테이블 추가

dye0p committed:
feat: 예약시간을 저장하는 API 구현

- 예약 시간을 시간 db에 추가할 수 있다.

dye0p committed:
refactor: 패키지 분리

dye0p committed:
refactor: 예약 시간 저장 후 TimeResponse를 반환하도록 변경

dye0p committed:
feat: 예약 시간을 조회하는 API 구현

dye0p committed:
feat: 특정 예약시간 번호의 예약시간을 삭제하는 API 구현

dye0p committed:
refactor: 예약시간 엔티티 객체 네이밍 변경

- Time -> ReservationTime

dye0p committed:
refactor: 예약 페이지 경로 변경

dye0p committed:
refactor: 테이블 스키마 재정의

- reservation 테이블의 외래키로 reservation_time 테이블의 pk 지정

dye0p committed:
refactor: 예약시 미리 추가한 예약 시간으로만 예약 할 수 있도록 변경

dye0p committed:
test: 중복 테스트 제거 및 디스플레이 네이밍 변경

dye0p committed:
refactor: 계층화 리팩터링

- Layered Architecture 적용

dye0p committed:
refactor: 지역변수 final 키워드 추가

dye0p committed:
refactor: 메서드 네이밍 명확하게 변경

dye0p committed:
refactor: dto 클래스 service 패키지로 이동

dye0p committed:
refactor: 클래스 레벨에 final 키워드 추가

dye0p committed:
refactor: 정적 팩터리 메서드 네이밍 명확하게 변경

- id가 없이 생성한다는 것을 직관적으로 표현하기 위해서 메서드 네이밍 변경

dye0p committed:
refactor: dto 변환 로직 service 로직으로 이동

dye0p commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,42 @@
+        this.time = time;
+    }
+
+    public static Reservation from(final Long id, final String name, final LocalDate date, final ReservationTime time) {
+        return new Reservation(id, name, date, time);
+    }
+
+    public static Reservation fromWithoutId(final String name, final LocalDate date, final ReservationTime time) {
+        return new Reservation(null, name, date, time);
```
db에 영속화 되기 전에는 Reservation에 id값이 null 입니다.
그래서 요청 dto를 엔티티 객체로 변환 하는 단계에서 id가 null인 객체를 생성해야 합니다.

처음엔 필드에 null을 넣는다는 것이 찝찝했습니다. 하지만 생각을 해봤을 때 영속화 되기 전이라면 어쩔 수 없이 id는 null일 수 밖에 없습니다. 그래서 id를 매개변수로 받지않는 정적 팩터리 메서드를 통해서 외부에서 null을 넣을 수 없도록 제한 하였습니다.

하지만 RowMapper로 매핑되는 과정에서 id를 외부에서 받아 생성해야 했기에 id를 받는 정적 팩터리 메서드를 정의하였고,
네이밍으로 id의 null여부를 구분해주었습니다.

이러한 방식에 대한 지노의 의견이 궁금합니다!

hyunssooo replied:
id를 db에서 받아오는 경우 null을 넣는 것은 어쩔 수 없고 id가 없는 상태이기 때문에 null 도 저는 어색하지는 않다고 생각해요. 

정적 팩터리 메서드를 사용한 점 좋습니다. 개인적으로는 생성자로 충분할 것 같다고 생각해요.

dye0p replied:
정적 팩터리 메서드를 사용한 이유는 `네이밍으로 생성자를 구분할 수 있다` 입니다! 

생성자로 충분하다고 생각하신 이유가 궁금합니다!
또한, 지노는 어떤 기준으로 정적 팩터리 메서드를 사용하시나요?

hyunssooo replied:
이 부분은 개인적인 기준인데요. 별도의 연산이 들어가지 않고 필드에 할당되는 경우는 생성자를 최대한 사용합니다. 생성자의 인자로 충분하다고 생각해서요!

dye0p replied:
그렇게 기준을 세울 수도 있겠네요! 좋은 의견 감사합니다 : )

hyunssooo commented on src/main/java/roomescape/repository/ReservationDao.java:
```diff
@@ -0,0 +1,67 @@
+                + "    t.start_at as time_value \n"
+                + "FROM reservation as r \n"
+                + "inner join reservation_time as t \n"
+                + "on r.time_id = t.id";
```
Text Blocks을 사용해 보는 것은 어떨까요?

dye0p replied:
문자열이 `+` 기호로 연결되어 있어서 가독성이 좋지 않다고 느꼈는데 새로운 기능을 알게 되었네요! 감사합니다 : )

[refactor: 긴 문자열 Text Blocks을 사용하여 가독성 향상](https://github.com/woowacourse/spring-roomescape-admin/pull/333/commits/da9d4febf8a83d7225ef91545fb93eb7792bc063)

hyunssooo commented on src/main/java/roomescape/repository/ReservationTimeDao.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public ReservationTime findById(final Long id) {
+        final String sql = "select id, start_at from reservation_time where id =?";
```
read와 find는 어떻게 다를까요?

dye0p replied:
처음 코드를 작성할 땐 `데이터를 읽는다.`에 초점을 맞추어 생각했습니다. 따라서 `read` 라는 네이밍을 사용하였습니다.
하지만 피드백을 통해서 찾아본 결과 `read`와 `find`간의 차이를 알게 되었습니다.

### read

- 기존의 데이터를 가져온다.
- 데이터가 확실하게 존재한다는 전제 하에 데이터를 가져오는 의미

### find

- 특정 조건에 맞는 데이터를 조회한다.
- 조건이 있을 수도 있고, 결과가 없을 수도 있음

`예약 목록을 조회한다` 라는 행동의 의미를 생각했을 때 `예약 목록이 없을 땐 빈 리스트를 반환`할 수 있어야 합니다.
따라서 read 보단 find가 더 적합하다고 판단하였습니다!

또한, 추후에 JPA를 사용하겠지만 JPA의 기본 메서드 네이밍도 findAll 이라고 정의해놓만큼 find가 표준에 가깝다고 생각합니다.

추가로 delete 메서드의 네이밍도 명확하게 변경해주었습니다!

네이밍에 대한 고민이 있었는데 남겨주신 피드백으로 짚고 넘어갈 수 있었습니다! 감사합니다 : )

[refactor: 데이터 접근 메서드 네이밍 명확하게 변경(read -> find)](https://github.com/woowacourse/spring-roomescape-admin/pull/333/commits/f5a6216235783e83a7dc9ffcd24b28ed4125c955)
[refactor: 데이터 삭제 메서드 네이밍 명확하게 변경(delete -> deleteById)](https://github.com/woowacourse/spring-roomescape-admin/pull/333/commits/f6fa2ad2ce3bf586845d2a267cd5aefd0f55bcd8)

hyunssooo commented on src/main/java/roomescape/service/dto/reservation/request/ReservationRequest.java:
```diff
@@ -0,0 +1,16 @@
+package roomescape.service.dto.reservation.request;
+
```
> Dto 의 패키지 위치?

네 적절한 이유로 괜찮은 선택이라고 생각합니다. 🙂 구성에 따라 다르긴 하지만 controller 의 dto(request, response) 와 service의 dto를 나눠 사용하기도 합니다. 개인적으로 엄격하게 구분하자면 나눠서 사용하는 것이 맞다고 생각해요. 하지만 말씀주신 이유로 트레이드 오프도 저는 충분히 좋다고 생각합니다. :)

dye0p replied:
리팩터링을 하는 과정에서 요청 dto를 controller 패키지로 옮기게 되었습니다. `요청 dto는 클라이언트의 요청을 가장 먼저 받는 controller에서 가장 먼저 사용된다`라고 생각했기 때문 입니다.

이러한 과정에서 2가지 문제가 발생했습니다.

## 문제

### 계층간 의존성 역전
`controller` 계층에 위치한 dto를 `service` 계층에서 사용하기 때문에 `service` 계층이 `controller` 계층을 의존하게 됩니다.
``` java
import org.springframework.stereotype.Service;
import roomescape.controller.reservation.request.ReservationRequest;
```
Layered Architecture 구조에서 의존성의 방향은 `상위 -> 하위` 처럼 `위에서 아래`로 흘러야 합니다. 위와 같은 상황은 의존성이 역전되어 의존성이 `아래에서 위`로 흐르게 됩니다. 이렇게 되버리면 각각의 계층이 독립적일 수 없습니다.

### 하나의 dto를 여러 계층에서 사용
하나의 `요청 dto`를 `controller` 와 `service` 계층에서 사용하게 됩니다. `dto`가 변경되게 된다면  `controller`와 `service` 까지 변경의 영향을 받게 될 것 입니다.

## 해결
말씀해주신 피드백에서 힌트를 얻어 문제를 개선할 수 있었습니다!

[refactor: controller dto와 service dto 분리](https://github.com/woowacourse/spring-roomescape-admin/pull/333/commits/dae724df9fdad0010208b32f44a06b420791f993)

## controller 와 service의 요청 dto를 분리
`controller` 와 `service`의 `요청 dto`를 각각 두었습니다.

controller 
``` java
    public ResponseEntity<ReservationResponse> save(@RequestBody ReservationRequest request) {
        final ReservationResponse response = reservationService.save(request.toServiceRequest());
```
service
``` java
   public ReservationResponse save(final ReservationServiceRequest reservationRequest) {
```
이렇게 `요청 dto`를 각각의 계층의 dto로 분리함으로써 계층간 의존성 문제를 해결할 수 있었습니다.

``` java
import org.springframework.stereotype.Service;
import roomescape.service.reservation.request.ReservationServiceRequest;
```

## 고민
### 왜 응답 dto는 분리하지 않았나?

`요청 dto`를 분리한 이유는 `service` 계층이 `controller` 계층에 의존하지 않도록 하기 위함이었습니다.
하지만 `응답 dto`의 경우는 `service`의 dto를 `controller`가 사용하기 때문에 계층간 의존성이 문제가 되지 않는다고 판단하였습니다.
(상위 계층은 하위 계층을 호출해도 괜찮기 때문)

지노는 `응답 dto`도 별도로 분리하는것이 좋다고 생각하시나요? 지노의 생각은 어떠신가요?

hyunssooo reviewed (changes requested):
안녕하세요. 도기! 
미션 잘 진행해 주셨습니다! 
몇 가지 가벼운 코멘트와 질문에 대한 제 생각 조금 남겼어요! 
확인 부탁드려요!

dye0p committed:
refactor: 긴 문자열 Text Blocks을 사용하여 가독성 향상

dye0p committed:
refactor: controller dto와 service dto 분리

- controller의 dto를 service에서 사용하게 되었을 때의 의존성 문제를 각각의 layer의 dto를 생성하여 의존관계를 끊음

- controller 패키지의 요청 dto를 service에서 사용하게 되었을 때 하위 계층인 service 계층이 상위 계층인 controller 계층을 의존하는 상황을 개선함

dye0p committed:
refactor: 데이터 접근 메서드 네이밍 명확하게 변경(read -> find)

- 기존의 데이터를 찾는 느낌의 read 보단, 조건에 따라서 데이터를 찾는 find가 더 적합함

dye0p committed:
refactor: 데이터 삭제 메서드 네이밍 명확하게 변경(delete -> deleteById)

- 특정한 id로 데이터를 삭제한다는 의미를 명확하게 함

dye0p reviewed (commented):
안녕하세요 지노! 말씀해주신 피드백 반영과 제 생각을 남겼습니다!

hyunssooo commented on src/main/java/roomescape/controller/reservation/ReservationController.java:
```diff
@@ -0,0 +1,45 @@
+package roomescape.controller.reservation;
+
```
controller 레이어 내부에서 reservation, reservationTime을 나눠주셨네요! 어떤 기준으로 나눠주셨나요? 어떤 의도인지 궁금해서요!

dye0p replied:
2가지 이유가 있었습니다! 

- 컨트롤러에서도 각각의 도메인에 대한 개념을 분리하기 위함
- 각 도메인에 해당하는 요청 dto가 존재하므로, 관련 컨트롤러와 dto를 함께 묶어서 패키지로 개념을 표현하고자 함

추가로 각각의 controller에 대해서 dto가 추가된다면 패키지를 나누었을 때 더욱 더 직관적으로 구분할 수 있을 거라고 생각이 들었습니다.

이에 대한 지노의 의견은 어떠신가요?

hyunssooo commented on src/main/java/roomescape/controller/reservation/request/ReservationRequest.java:
```diff
@@ -0,0 +1,14 @@
+package roomescape.controller.reservation.request;
+
```
https://github.com/woowacourse/spring-roomescape-admin/pull/333#discussion_r2062600326

저라면 나누기로 했으면 나누고 합치기로 했으면 합칠 것 같습니다. 지금은 요청은 나누고 응답은 합쳐서 일관성이 좀 없어 보이긴 하는데요. dto는 해당 레이어의 인터페이스로 볼 수 있을 것 같아요. controller의 Request, Response는 사용하는 클라이언트에게 제공될 인터페이스라고 볼 수 있을 것 같아요. 
그리고 모듈을 나눠서도 많이 사용하는데 이런 경우는 아래에서 위로 의존성 방향이 흐를 수 없습니다. 그렇다면 저는 application layer에 Request, Response dto를 위치하게 하지 않게 되더라고요. (application layer를 사용하는 모듈이 api만 있는 것이 아님)

dye0p replied:
저도 지노의 의견에 동의합니다! 하지만 분리를 시도하면서 service의 response를 controller로 넘길 때 매핑을 해줘야 하는데, 이것을 service에서 하게 되면 controller 계층을 알게 되고, controller 에서 해주게 되면 controller에 service의 응답 dto가 노출됩니다. 이것의 딜레마에 빠져 수정을 하지 못하였습니다.

현재 구조에서는 어쩔 수 없이 응답 dto를 두 계층에서 노출해야 하는데 지노는 어떻게 생각하시나요?

### 모듈화 구조

추가로 모듈 구조에 대해서 공부할 수 있었습니다! 
모듈로 구조화를 하게되면 각각의 독립적인 구조로 개발했을때의 장점은

- 관심사의 분리가 명확해짐
- 독립적인 테스트가 가능해짐
- 의존성의 방향이 명확해짐

위와 같다고 생각이 들었습니다! 

아직 말씀해주신 모듈화에 대해서 미흡하게 알고 있는 것 같아요! 혹시 제가 잘못 알고 있거나, 추가로 학습하면 좋을 키워드가 있다면 피드백 부탁드립니다 : )

hyunssooo replied:
현재 구조에서는 굳이 dto를 나눠쓰지 않을 것 같아요. application layer에 dto 를 위치하고 요청, 응답, service에서 다 같이 사용할 것 같습니다. 나눠써도 좋지만 관리해야할 파일이 많아지고 응답과 service의 리턴값이 다르지 않기 때문에 트레이드 오프를 했다고 생각해요. 
예를 들어 ReservationAdminController, ReservationController 가 별도로 존재하고 두 클래스 모두 ReservationService를 의존하고 있다면 Dto를 나눠볼 것 을 고려해 볼 수 있을 것 같아요.

hyunssooo commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -1,28 +1,37 @@
-                .then().log().all()
-                .statusCode(200);
-    }
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+    @Autowired
+    private ReservationTimeController reservationTimeController;
+    @Autowired
+    private ReservationController reservationController;
+
```
테스트는 각 레이어별로 나눠서 작성해 보는 것은 어떨까요??

dye0p replied:
지금은 하나의 테스트에 위치하기 때문에 알아보기 힘들다고 생각합니다. 때문에 다음과 같이 리팩토링 하였습니다!

- `service` 테스트 추가
- `dao` 테스트 추가
- 기존의 API 테스트를 각각의  `Controller` 테스트에 이동

dao 테스트 중 한가지 고민이 있었습니다.

### 테스트를 위한 `setter` 
테스트를 위한 FakeDao를 구현하던 중 `Reservation`과 `ReservationTime` 객체에 id를 주입해주어야 하는 상황이 있었습니다.
그래서 어쩔 수 없이 `setter`를 열어주었지만 현재로서는 `Junit` 테스트로만 제한했기 때문에 어느정도 납득할 수 있었지만, 기존에 `final`로 사용되던 `id` 필드를 테스트를 위해서 `setter`를 열어주는 것이 적절한지 고민이 되었습니다. 

``` java
    public long save(final Reservation reservation) {
        reservation.setId(id);
        reservations.add(reservation);
        return reservation.getId();
    }
```

이에 대해 지노는 어떻게 생각하시나요?

hyunssooo replied:
테스트를 위한 코드가 프로덕션 코드에 존재해도 괜찮을까요?

hyunssooo reviewed (changes requested):
안녕하세요. 도기!
남겨주신 코멘트 확인했어요! 
몇 가지 제 생각 남겼는데요! 확인 부탁드려요!

dye0p committed:
test: DAO 테스트 추가

- 테스트를 위해서 dao구현체를 추상화 하였음
- Junit으로 테스트하기 위해 테스트를 위한 FakeDAO 객체 생성

dye0p committed:
test: service 객체 테스트 추가

- 기존 구현체의 구체 타입을 의존하던 service 필드 객체의 타입을 인터페이스 타입을 의존하도록 변경하여 테스트를 가능하게 함

dye0p committed:
test: controller 테스트 추가

- 기존의 MissionStepTest 의 API 테스트를 각각의 controller 테스트로 이동

dye0p reviewed (commented):
안녕하세요 지노! 남겨주신 피드백과 함께 저의 생각을 남겼습니다! 항상 좋은 피드백 감사합니다 : )

hyunssooo commented on src/main/java/roomescape/repository/reservation/JdbcReservationDao.java:
```diff
@@ -0,0 +1,72 @@
+package roomescape.repository.reservation;
+
```
https://github.com/woowacourse/spring-roomescape-admin/pull/333#discussion_r2065288616

> 컨트롤러에서도 각각의 도메인에 대한 개념을 분리하기 위함

개념적으로 나눠주신 부분은 충분히 이해가 됩니다. 다만 예약, 예약 시간은 별도의 개념일까요? 예약은 예약 시간이 필수고 예약 시간은 예약이 없으면 안 되는 개념인 것 같아서요. 이 경우 패키지까지 분리해야 할까? 두 개념은 별도의 개념이 맞을까? 라는 생각이 들어서요. 이 부분은 정답은 없지만 충분히 고민해 볼 만한 주제라고 생각합니다. :) 이 부분은 레벨 2 진행하시면서 조금씩 생각이 변할 수 도 있을 것 같아요. 도기만의 기준을 찾으시면 좋을 것 같아요.

hyunssooo reviewed (approved):
안녕하세요. 도기! 
이번 미션 진행하신다고 고생 많으셨습니다! 
질문 주신 부분에 제 생각 남겼어요! 확인 부탁드립니다! 
궁금하시거나 함께 고민이 필요한 부분이 있으시면 언제든 편하게 연락주세요!
다음 미션도 잘 진행하시길 바랄게요!
